dipy/denoise/localpca.py

```diff
@@ -0,0 +1,118 @@
+    """
+    if not arr.ndim == 4:
+        raise ValueError("PCA denoising can only be performed on 4D arrays.",
+                         arr.shape)
+
+    patch_size = 2 * patch_radius + 1
+
+    if patch_size ** 3 < arr.shape[-1]:
+        e_s = "You asked for PCA denoising with a "
+        e_s += "patch_radius of {0} ".format(patch_radius)
+        e_s += "for data with {0} directions. ".format(arr.shape[-1])
+        e_s += "This would result in an ill-conditioned PCA matrix. "
+        e_s += "Please increase the patch_radius."
+        raise ValueError(e_s)
+
+    if isinstance(sigma, np.ndarray):
+        if not sigma.shape == arr.shape[:-1]:
+            e_s = "You provided a sigma array with a shape"
+            e_s += "{0} for data with".format(sigma.shape)
+            e_s += "shape {0}. Please provide a sigma array".format(arr.shape)
+            e_s += " that matches the spatial dimensions of the data."
+            raise ValueError(e_s)
+
+    tau = np.ones(arr.shape[:-1]) * ((tau_factor * sigma) ** 2)
+    del sigma
+    # declare arrays for theta and thetax
+    theta = np.zeros(arr.shape, dtype=np.float64)
+    thetax = np.zeros(arr.shape, dtype=np.float64)
+
+    # loop around and find the 3D patch for each direction at each pixel
+    for k in range(patch_radius, arr.shape[2] - patch_radius):
+        for j in range(patch_radius, arr.shape[1] - patch_radius):
+            for i in range(patch_radius, arr.shape[0] - patch_radius):
+                # Shorthand for indexing variables:
+                ix1 = i - patch_radius
+                ix2 = i + patch_radius + 1
+                jx1 = j - patch_radius
+                jx2 = j + patch_radius + 1
+                kx1 = k - patch_radius
+                kx2 = k + patch_radius + 1
+
+                X = arr[ix1:ix2, jx1:jx2, kx1:kx2].reshape(
+                                patch_size ** 3, arr.shape[-1])
+                # compute the mean and normalize
+                M = np.mean(X, axis=0)
+                X = X - M
+                # PCA using an SVD
+                U, S, Vt = svd(X, *svd_args)[:3]
+                # Items in S are the eigenvalues, but in ascending order
+                # We invert the order (=> descending), square and normalize
+                # \lambda_i = s_i^2 / n
+                d = S[::-1] ** 2 / X.shape[0]
+                # Rows of Vt are eigenvectors, but also in ascending eigenvalue
+                # order:
+                W = Vt[::-1].T
+                # Threshold by tau:
+                W[:, d < tau[i, j, k]] = 0
+                # This is equations 1 and 2 in Manjon 2013:
+                Xest = X.dot(W).dot(W.T) + M
+                Xest = Xest.reshape(patch_size,
+                                    patch_size,
+                                    patch_size, arr.shape[-1])
+                # This is equation 3 in Manjon 2013:
+                theta[ix1:ix2, jx1:jx2, kx1:kx2] += 1.0 / (1.0 + np.sum(d > 0))
```

Comment:
Honestly, I really just don't see how to do this. Since this doesn't change the functionality, I suggest that you do another PR with this improvement once we get this one merged (or you can do another PR against this branch right now, if you want).